src/coreclr/src/jit/codegencommon.cpp

```diff
@@ -6308,71 +6308,165 @@ void CodeGen::genZeroInitFrame(int untrLclHi, int untrLclLo, regNumber initReg,
-            regSet.verifyRegUsed(REG_ESI);
-        }
+            // For register arguments we may have to save ECX
+            if (intRegState.rsCalleeRegArgMaskLiveIn & RBM_ECX)
+            {
+                noway_assert(regSet.rsRegsModified(RBM_ESI));
+                inst_RV_RV(INS_mov, REG_ESI, REG_ECX);
+                regSet.verifyRegUsed(REG_ESI);
+            }
 #endif // !UNIX_AMD64_ABI
-
-        noway_assert((intRegState.rsCalleeRegArgMaskLiveIn & RBM_EAX) == 0);
-
-        GetEmitter()->emitIns_R_AR(INS_lea, EA_PTRSIZE, REG_EDI, genFramePointerReg(), untrLclLo);
-        regSet.verifyRegUsed(REG_EDI);
-
-        inst_RV_IV(INS_mov, REG_ECX, (untrLclHi - untrLclLo) / sizeof(int), EA_4BYTE);
-        instGen_Set_Reg_To_Zero(EA_PTRSIZE, REG_EAX);
-        instGen(INS_r_stosd);
+            // We need to 32 byte align the rep stosd as there is a penality if it is not 32 byte aligned
+            // Grab a non-argument, non-callee saved XMM reg
+            emitter*  emit     = GetEmitter();
+            regNumber frameReg = genFramePointerReg();
+            // Get block start
+            emit->emitIns_R_AR(INS_lea, EA_PTRSIZE, REG_EDI, frameReg, untrLclLo);
+#ifdef UNIX_AMD64_ABI
+            // System V x64 first temp reg is xmm8
+            regNumber zeroSIMDReg = genRegNumFromMask(RBM_XMM8);
+#else  // !UNIX_AMD64_ABI
+            // Windows first temp reg is xmm4
+            regNumber zeroSIMDReg = genRegNumFromMask(RBM_XMM4);
+#endif // UNIX_AMD64_ABI
+            // Zero xmm reg
+            emit->emitIns_R_R(INS_xorps, EA_16BYTE, zeroSIMDReg, zeroSIMDReg);
+            // Zero first 32 bytes (may overlap with next 32 byte alignment)
+            emit->emitIns_AR_R(INS_movdqu, EA_ATTR(XMM_REGSIZE_BYTES), zeroSIMDReg, frameReg, untrLclLo);
+            emit->emitIns_AR_R(INS_movdqu, EA_ATTR(XMM_REGSIZE_BYTES), zeroSIMDReg, frameReg,
+                               untrLclLo + XMM_REGSIZE_BYTES);
+            // Stash current block start
+            emit->emitIns_R_R(INS_mov, EA_PTRSIZE, REG_ECX, REG_EDI);
+            // Add 32 to hit or go over the next alignment
+            emit->emitIns_R_I(INS_add, EA_PTRSIZE, REG_EDI, 32);
+            // Clear all the unaligned bits, RDI now contains the aligned address
+            emit->emitIns_R_I(INS_and, EA_PTRSIZE, REG_EDI, -32);
+            // Copy the aligned address
+            emit->emitIns_R_R(INS_mov, EA_PTRSIZE, REG_EAX, REG_EDI);
+            // Subtract the unaligned address from aligned to get amount to subtract from count
+            emit->emitIns_R_R(INS_sub, EA_PTRSIZE, REG_EAX, REG_ECX);
+            // Output count of bytes to clear
+            inst_RV_IV(INS_mov, REG_ECX, blkSize, EA_PTRSIZE);
+            // Subtract the unaligned already cleared
+            emit->emitIns_R_R(INS_sub, EA_PTRSIZE, REG_ECX, REG_EAX);
+
+            instGen_Set_Reg_To_Zero(EA_PTRSIZE, REG_EAX);
+            instGen(INS_r_stosb);
```

Comment:
The last time stosd->stosb change was tested it turned out to cause significant regressions on AMD processors.
Also the comments above all still mention stosd.

Reply:
So `stosq` would be better it looks like

Reply:
I think its a case where Intel has better optimized `stosb` (and on newer machines `stosb` will perform better than `stosd`) but on AMD, they recommend using the largest properly aligned data size (so `stosd`/`stosq` will perform better).

For reference, the AMD Optimization manuals can be found here: <https://developer.amd.com/resources/developer-guides-manuals/>
You probably want the `Family 15h Processors` manual which has `9.3 Repeated String Instructions` (the 17h manuals just have some Zen specific additionals and require 15h for a complete view).

Reply:
Changed to using highest stride version